src/default_storage.jl

```diff
@@ -0,0 +1,393 @@
+    _TopLevelRuntimeWithStorage, _TracingRuntimeWithStorage
+
+import ..Salsa.Debug: @debug_mode, @dbg_log_trace
+
+const Revision = Int
+
+struct InputValue{T}
+    value::T
+    changed_at::Revision
+    # Allow converting from abitrary values (e.g. creating a Vector{String} from []).
+    InputValue{T}(v, changed_at) where {T} = new{T}(v, changed_at)
+end
+InputValue(v::T, changed_at) where {T} = InputValue{T}(v, changed_at)
+
+_changed_at(v::InputValue)::Revision = v.changed_at
+
+# This struct is mutable so that we can edit the Revisions in-place without having to
+# reallocate a new DerivedValue. This is a performance optimization to avoid allocations.
+# Since this struct contains a Vector, it will not be isbits, so it will be heap-allocated
+# anyway (though this may change in the future, with Julia's stack allocation patch in 1.5;
+# consider re-evaluating this decision then).
+mutable struct DerivedValue{T}
+    value::T
+    # A list of all the computations that were accessed directly from this derived function
+    # (not all recursive dependencies) when computing this derived value.
+    dependencies::Vector{DependencyKey}
+    # These Revisions are used to determine whether we need to re-compute this value or not.
+    # We need both of them in order to correctly implement the Early-Exit Optimization.
+    changed_at::Revision
+    verified_at::Revision
+end
+
+_changed_at(v::DerivedValue)::Revision = v.changed_at
+
+
+
+const InputMapType = IdDict{InputKey,Dict}
+const DerivedFunctionMapType = IdDict{DerivedKey,Dict}
+
+mutable struct DefaultStorage <: AbstractSalsaStorage
+    # The entire Salsa storage is protected by this lock. All accesses and
+    # modifications to the storage should be made within this lock.
+    lock::Base.ReentrantLock
+
+    # This is bumped every time the storage changes at all, in `set_input!` and
+    # `delete_input!`. Whenever this is bumped, we have to check every derived funciton the
+    # first time it is called to see if it is still valid.
+    # NOTE: This is guaranteed not to change while any derived functions are currently
+    # running anywhere on any threads, which we ensure by tracking the count of active
+    # derived functions via `derived_functions_active`, below.
+    current_revision::Int64
+
+    # We use one big dictionary for the inputs, storing them all together, to reduce
+    # allocating a new dictionary for every input.
+    # TODO: Do more performance investigation for the tradeoff between sharing a dictionary
+    # vs having separate dicts per method. It seems like the current decision is exactly
+    # opposite of what would be best: Since DerivedValues are not isbits, they will always
+    # be heap allocated, so there's no reason to strongly type their dict. But inputs can
+    # be isbits, so it's probably worth specailizing them.
+    inputs_map::Dict{Tuple{Type,Tuple},InputValue}
+    derived_function_maps::DerivedFunctionMapType
```

Comment:
I've opened https://github.com/RelationalAI-oss/Salsa.jl/issues/18 for this TODO